### ISSUE TITLE
`react/no-unstable-nested-components` does not allow ignoring only object properties

### ISSUE THREAD

**Issue by bertho-zero** [jsx-eslint/eslint-plugin-react#3482]
I would like to keep this rule active but disable it for object properties, I have this:

```jsx
function MyComponent() {
  return (
    <p>
      {intl.formatMessage(messages.allSelected, {
        size: selectedEvents.size,
        b: chunks => <b>{chunks}</b>, // <-- report error here without `allowAsProps`
      })}
    </p>
  );
}
```
But adding the `allowAsProps: true` option allows me to do things like this:
```jsx
<Dropdown
  Trigger={({ onClick }) => ( // <-- don't report error here with `allowAsProps`
    <button type="button" onClick={onClick}>
      Bla bla
      <span className="caret" />
    </button>
  )}
>
  {/* ... */}
</Dropdown>
```

**Comment by ljharb** [comment]
I'm a bit confused. `Trigger` here is definitely a component, and should be warned on - but `b` isn't a component, because it's got a lowercase name (if it was `B` i'd say it *is* a component).
